server/src/main.rs

```diff
@@ -0,0 +1,156 @@
+            let res = entire_body.map( |body| {
+                // Extract X-Original-URI header value, panic for no header
+                let req_uri = match parts.headers.get(&*X_ORIGINAL_URI) {
+                    Some(s) => s.to_str().unwrap(),
+                    None => panic!("Request does not contain \"X-Original-URI\" header."),
+                };
+                let body_string = String::from_utf8(body.to_vec()).expect("Found invalid UTF-8");
+                let payload = create_payload(parts.method.to_string(), req_uri.to_string(), body_string);
+                let is_verified = verify_request(payload, parts.headers, &HP_PUBLIC_KEY);
+                respond_success(is_verified)
+            });
+
+            Either::A(res)
+        }
+        _ => {
+            let res = future::ok(respond_success(false));
+            Either::B(res)
+        }
+    }
+}
+
+fn create_payload (method: String, uri: String, body_string: String) -> String {
+    let d = json!({
+        "method": method.to_lowercase(), // make sure verb is to lowercase
+        "uri": uri,
+        "body": body_string
+    }); 
+
+    // Serialize it to a JSON string.
+    serde_json::to_string(&d).unwrap()
+}
+
+fn verify_request(payload: String, headers: HeaderMap<HeaderValue>, public_key: &PublicKey) -> bool {
+    if let Some(signature_base64) = headers.get(&*X_HPOS_ADMIN_SIGNATURE) {
+        if let Ok(signature_vec) = decode_config(&signature_base64, base64::STANDARD_NO_PAD) {
+            if let Ok(signature_bytes) = Signature::from_bytes(&signature_vec) {
+                if public_key.verify(&payload.as_bytes(), &signature_bytes).is_ok() {
+                    return true
+                }
+            }
+        }
+    }
+    return false
+}
+
+fn respond_success (is_verified: bool) -> hyper::Response<Body> {
+    // construct response based on verification status
+    match is_verified {
+        true => {
+            Response::builder()
+                .status(StatusCode::OK)
+                .body(Body::empty())
+                .unwrap()
+        },
+        _ => {
+            Response::builder()
+                .status(StatusCode::UNAUTHORIZED)
+                .body(Body::empty())
+                .unwrap()
+        }
+    }
+}
+
+fn read_hp_pubkey() -> PublicKey {
+    let hpos_state_path = env::var("HPOS_STATE_PATH").expect("HPOS_STATE_PATH environmental variable is not present");
+
+    println!("Reading HP Admin Public Key from {}.", hpos_state_path);
+
+    // Read from path
+    let contents = fs::read(hpos_state_path)
+        .expect("Something went wrong reading HP Admin Public Key from file");
+
+    // Parse content
+    let hpos_state: State = serde_json::from_slice(&contents).unwrap();
+    hpos_state.get_admin_public_key().expect("HP Admin Public key seems to be corrupted")
+}
+
+fn main() {
+    // Listen on http socket port 2884 - "auth" in phonespell
+    let listen_address = ([127,0,0,1], 2884).into();
```

Comment:
Let's use Unix sockets for new services.

Reply:
I am implementing http according to our conversation. As a reminder- accessing this service is not any usable attack vector. It allows though to finish this task in the reasonable timeframe.